src/scanForInputs.js

```diff
@@ -2,24 +2,88 @@ const {Form} = require('./Form/Form')
 const {notifyWebApp} = require('./autofill-utils')
 const {SUBMIT_BUTTON_SELECTOR, FORM_INPUTS_SELECTOR} = require('./Form/selectors-css')
 
-/** @type Map<HTMLElement, Form> */
-const _forms = new Map()
-
 /**
- * This will return `init` and `findEligibleInputs` which allows consumers
- * to either `init` if in the context of a webpage, or alternatively just perform
- * the synchronous mutations via findEligibleInputs
- *
- * @param DeviceInterface
- * @param {Map<HTMLElement, Form>} [forms]
- * @returns {{
- *   init: () => () => void,
- *   findEligibleInputs: (element: Element|Document) => void
- * }}
+ * This allows:
+ *   1) synchronous DOM scanning + mutations - via `new Scanner(device).findEligibleInputs(document)`
+ *   2) or, as above + a debounced mutation observer to re-run the scan after the given time
  */
-const scanForInputs = (DeviceInterface, forms = _forms) => {
-    const getParentForm = (input) => {
-        if (input.form) return input.form
+class Scanner {
+    /** @type Map<HTMLElement, Form> */
+    forms;
+    /** @type {number} How long should we wait for the DOM to stop changing before we attempt to re-scan? */
+    debounceTimePeriod = 500
+    /** @type {any|undefined} the timer to reset */
+    debounceTimer;
+    /** @type {Set<HTMLElement|Document>} stored changed elements until they can be processed */
+    changedElements = new Set()
+
+    /**
+     * @param {import("./DeviceInterface/InterfacePrototype")} device
+     * @param {Map<HTMLElement, Form>} forms
+     */
+    constructor (device, forms) {
+        this.device = device
+        this.forms = forms
+    }
+    /**
+     * Call this to scan once and then watch for changes.
+     *
+     * Call the returned function to remove listeners.
+     * @returns {() => void}
+     */
+    init () {
+        const delay = this.device.initialSetupDelayMs
+
+        // if the delay is zero, (chrome/firefox etc) then use `requestIdleCallback`
+        if (delay === 0) {
+            window.requestIdleCallback(() => this.scanAndObserve())
+        } else {
+            // otherwise, use the delay time to defer the initial scan
+            setTimeout(() => this.scanAndObserve(), this.device.initialSetupDelayMs)
+        }
+        return () => {
+            // remove Dax, listeners, timers, and observers
+            clearTimeout(this.debounceTimer)
+            this.mutObs.disconnect()
+            this.forms.forEach(form => {
+                form.resetAllInputs()
+                form.removeAllDecorations()
+            })
+            this.forms.clear()
+            notifyWebApp({ deviceSignedIn: {value: false} })
+        }
+    }
+
+    /**
+     * Scan the page and begin observing changes
+     */
+    scanAndObserve () {
+        window.performance?.mark?.('findEligibleInputs:start')
```

Comment:
Shouldn't we run `window.performance` only for the tests?

Reply:
These markers are safe to leave in - they have no impact on production code, but are handy to be in place when needed